File: packages/php/tests/ResourcesTest.php
<?php

declare(strict_types=1);

namespace MissionInbox\Tests;

use PHPUnit\Framework\TestCase;

final class ResourcesTest extends TestCase
{
    public function testSendingIdentifiersList(): void
    {
        $captured = [];
        $mi = TestHelpers::newClient([TestHelpers::jsonResponse(200, [])], $captured);
        $mi->sendingIdentifiers->list();
        $this->assertSame('/api/sending-identifiers', $captured[0]['request']->getUri()->getPath());
        $this->assertSame('GET', $captured[0]['request']->getMethod());
    }

    public function testSendingIdentifiersUpdatePatchesBody(): void
    {
        $captured = [];
        $mi = TestHelpers::newClient([TestHelpers::jsonResponse(200, ['id' => 'x'])], $captured);
        $mi->sendingIdentifiers->update('uuid-1', ['displayName' => 'Acme']);
        $this->assertSame('PATCH', $captured[0]['request']->getMethod());
        $this->assertSame(['displayName' => 'Acme'], TestHelpers::bodyOf($captured[0]['request']));
    }

    public function testDomainsListForwardsQueryString(): void
    {
        $captured = [];
        $mi = TestHelpers::newClient([TestHelpers::jsonResponse(200, ['data' => []])], $captured);
        $mi->domains->list(['verified' => true, 'limit' => 50, 'page' => 2]);
        $query = $captured[0]['request']->getUri()->getQuery();
        $this->assertStringContainsString('verified=true', $query);
        $this->assertStringContainsString('limit=50', $query);
        $this->assertStringContainsString('page=2', $query);
    }

    public function testDomainsGetByIdPathIncludesByIdSegment(): void
    {
        $captured = [];
        $mi = TestHelpers::newClient([TestHelpers::jsonResponse(200, ['id' => 'x'])], $captured);
        $mi->domains->get('uuid-1');
        $this->assertSame('/api/domains/by-id/uuid-1', $captured[0]['request']->getUri()->getPath());
    }

    public function testDomainsVerifyPostsDomainNameInBody(): void
    {
        $captured = [];
        $mi = TestHelpers::newClient(
            [TestHelpers::jsonResponse(200, ['fullyVerified' => true, 'dnsChecks' => ['dkim' => ['status' => 'OK']], 'message' => ''])],
            $captured,
        );
        $mi->domains->verify('acme.com');
        $this->assertSame('/api/domains/verify', $captured[0]['request']->getUri()->getPath());
        $this->assertSame(['domainName' => 'acme.com'], TestHelpers::bodyOf($captured[0]['request']));
    }

    public function testDomainsBulkCreateReturnsTaskId(): void
    {
        $mi = TestHelpers::newClient([TestHelpers::jsonResponse(200, ['taskId' => 't-1', 'message' => 'started'])]);
        $result = $mi->domains->bulkCreate(['domains' => [['domainName' => 'a.com']]]);
        $this->assertSame('t-1', $result['taskId']);
    }

    public function testDomainsExportCsvReturnsTextBody(): void
    {
        $mi = TestHelpers::newClient([TestHelpers::textResponse(200, "name,verified\nacme.com,true", 'text/csv')]);
        $csv = $mi->domains->exportCsv();
        $this->assertSame("name,verified\nacme.com,true", $csv);
    }

    public function testDomainsRedirectSetupPutsRedirectUrl(): void
    {
        $captured = [];
        $mi = TestHelpers::newClient(
            [TestHelpers::jsonResponse(200, ['success' => true, 'action' => 'created'])],
            $captured,
        );
        $mi->domains->redirects->setup('acme.com', ['redirectUrl' => 'https://www.acme.com', 'forceHttps' => true]);
        $this->assertSame('PUT', $captured[0]['request']->getMethod());
        $this->assertSame('/api/domains/acme.com/redirect', $captured[0]['request']->getUri()->getPath());
        $this->assertSame(
            ['redirectUrl' => 'https://www.acme.com', 'forceHttps' => true],
            TestHelpers::bodyOf($captured[0]['request']),
        );
    }

    public function testProjectsAssignDomainsPatch(): void
    {
        $captured = [];
        $mi = TestHelpers::newClient(
            [TestHelpers::jsonResponse(200, ['project' => [], 'total' => 1, 'successful' => 1, 'failed' => 0, 'reassigned' => 0, 'results' => []])],
            $captured,
        );
        $mi->projects->assignDomains('p-1', ['domainNames' => ['a.com']]);
        $this->assertSame('PATCH', $captured[0]['request']->getMethod());
        $this->assertSame('/api/projects/p-1/domains', $captured[0]['request']->getUri()->getPath());
    }

    public function testAnalyticsActivityGraphQueryString(): void
    {
        $captured = [];
        $mi = TestHelpers::newClient([TestHelpers::jsonResponse(200, [])], $captured);
        $mi->analytics->getActivityGraph([
            'granularity' => 'daily',
            'startDate' => '2026-01-01T00:00:00Z',
            'endDate' => '2026-01-31T23:59:59Z',
            'counters' => ['outgoing', 'bounces'],
        ]);
        $query = $captured[0]['request']->getUri()->getQuery();
        $this->assertStringContainsString('granularity=daily', $query);
        $this->assertStringContainsString('counters=outgoing', $query);
        $this->assertStringContainsString('counters=bounces', $query);
    }

    public function testEmailQueueRetryPostsToCorrectPath(): void
    {
        $captured = [];
        $mi = TestHelpers::newClient([TestHelpers::jsonResponse(200, [])], $captured);
        $mi->emailQueue->retry('q-1');
        $this->assertSame('POST', $captured[0]['request']->getMethod());
        $this->assertSame('/api/email/queue/q-1/retry', $captured[0]['request']->getUri()->getPath());
    }

    public function testHealthCheckReturnsTextBody(): void
    {
        $mi = TestHelpers::newClient([TestHelpers::textResponse(200, 'Healthy')]);
        $result = $mi->health->check();
        $this->assertSame('Healthy', $result);
    }
}

File: packages/java/src/test/java/com/missioninbox/ResourcesTest.java
package com.missioninbox;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

final class ResourcesTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private MissionInbox newClient(MockServer server) {
        return MissionInbox.builder()
                .apiKey("test-key")
                .baseUrl(server.baseUrl())
                .maxRetries(0)
                .timeout(Duration.ofSeconds(5))
                .build();
    }

    @Test
    void sending_identifiers_list_get() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueueJson(200, List.of());
            MissionInbox mi = newClient(server);
            mi.sendingIdentifiers.list();
            assertEquals("GET", server.lastRequest().method);
            assertEquals("/api/sending-identifiers", server.lastRequest().path);
        }
    }

    @Test
    void sending_identifiers_update_patches() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueueJson(200, Map.of("id", "x"));
            MissionInbox mi = newClient(server);
            mi.sendingIdentifiers.update("uuid-1", Map.of("displayName", "Acme"));
            assertEquals("PATCH", server.lastRequest().method);
            Map<String, Object> body = MAPPER.readValue(server.lastRequest().body, new TypeReference<Map<String, Object>>() {});
            assertEquals("Acme", body.get("displayName"));
        }
    }

    @Test
    void domains_list_query_string() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueueJson(200, Map.of("data", List.of()));
            MissionInbox mi = newClient(server);
            mi.domains.list(Map.of("verified", true, "limit", 50, "page", 2));
            String q = server.lastRequest().rawQuery;
            assertNotNull(q);
            assertTrue(q.contains("verified=true"));
            assertTrue(q.contains("limit=50"));
            assertTrue(q.contains("page=2"));
        }
    }

    @Test
    void domains_get_by_id_path() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueueJson(200, Map.of("id", "x"));
            MissionInbox mi = newClient(server);
            mi.domains.get("uuid-1");
            assertEquals("/api/domains/by-id/uuid-1", server.lastRequest().path);
        }
    }

    @Test
    void domains_verify() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueueJson(200, Map.of("fullyVerified", true, "dnsChecks", Map.of("dkim", Map.of("status", "OK")), "message", ""));
            MissionInbox mi = newClient(server);
            mi.domains.verify("acme.com");
            assertEquals("/api/domains/verify", server.lastRequest().path);
            Map<String, Object> body = MAPPER.readValue(server.lastRequest().body, new TypeReference<Map<String, Object>>() {});
            assertEquals("acme.com", body.get("domainName"));
        }
    }

    @Test
    void domains_bulk_create_returns_task_id() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueueJson(200, Map.of("taskId", "t-1", "message", "started"));
            MissionInbox mi = newClient(server);
            Map<String, Object> result = mi.domains.bulkCreate(List.of(Map.of("domainName", "a.com")));
            assertEquals("t-1", result.get("taskId"));
        }
    }

    @Test
    void domains_export_csv_returns_text() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue(200, "text/csv", "name,verified\nacme.com,true");
            MissionInbox mi = newClient(server);
            String csv = mi.domains.exportCsv();
            assertEquals("name,verified\nacme.com,true", csv);
        }
    }

    @Test
    void domain_redirect_setup_puts() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueueJson(200, Map.of("success", true, "action", "created"));
            MissionInbox mi = newClient(server);
            mi.domains.redirects.setup("acme.com", Map.of("redirectUrl", "https://www.acme.com", "forceHttps", true));
            assertEquals("PUT", server.lastRequest().method);
            assertEquals("/api/domains/acme.com/redirect", server.lastRequest().path);
        }
    }

    @Test
    void projects_assign_domains_patch() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueueJson(200, Map.of("project", Map.of(), "total", 1, "successful", 1, "failed", 0, "reassigned", 0, "results", List.of()));
            MissionInbox mi = newClient(server);
            mi.projects.assignDomains("p-1", List.of("a.com"));
            assertEquals("PATCH", server.lastRequest().method);
            assertEquals("/api/projects/p-1/domains", server.lastRequest().path);
        }
    }

    @Test
    void analytics_activity_graph_query() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueueJson(200, Map.of());
            MissionInbox mi = newClient(server);
            mi.analytics.getActivityGraph(
                    "daily",
                    "2026-01-01T00:00:00Z",
                    "2026-01-31T23:59:59Z",
                    List.of("outgoing", "bounces")
            );
            String q = server.lastRequest().rawQuery;
            assertTrue(q.contains("granularity=daily"));
            assertTrue(q.contains("counters=outgoing"));
            assertTrue(q.contains("counters=bounces"));
        }
    }

    @Test
    void email_queue_retry_post() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueueJson(200, Map.of());
            MissionInbox mi = newClient(server);
            mi.emailQueue.retry("q-1");
            assertEquals("POST", server.lastRequest().method);
            assertEquals("/api/email/queue/q-1/retry", server.lastRequest().path);
        }
    }

    @Test
    void health_check_returns_text() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue(200, "text/plain", "Healthy");
            MissionInbox mi = newClient(server);
            assertEquals("Healthy", mi.health.check());
        }
    }
}
